### PR TITLE
Use a cargo workspace

### LINKED ISSUE #411: Use Cargo Workspaces
This feature landed relatively recently and I _think_ it can benefit us by cutting down on test times. We should figure out how to use it and integrate it into our crates.

### PR DESCRIPTION
**Pull request by Bobo1239**
Fixes #411.

Uses a ["virtual" Cargo.toml](https://github.com/rust-lang/rfcs/blob/master/text/1525-cargo-workspace.md#virtual-cargotoml) as we don't have a root crate.

### CODE REVIEW

**Comment by sgrif** [comment]
Do we need to update each of the children to point to the workspace as well?

**Comment by killercup** [comment]
The compile-fail tests can't seem to find the necessary crates ([log](https://travis-ci.org/diesel-rs/diesel/jobs/153530965#L1135)).

**Comment by Bobo1239** [comment]
@sgrif That should not be necessary because of [implicit relations](https://github.com/rust-lang/rfcs/blob/master/text/1525-cargo-workspace.md#implicit-relations).

> The `package.workspace` can be omitted if it would only contain `../` (or some
> repetition of it). That is, if the root of a workspace is hierarchically the
> first `Cargo.toml` with `[workspace]` above a crate in the filesystem, then that
> crate can omit the `package.workspace` key.

File: .travis.yml [hunk 1]
```diff
@@ -14,7 +14,10 @@ before_script:
   - psql -c 'create database diesel_schema;' -U postgres
 script:
 - |
-  (cd diesel && travis-cargo doc -- --features "postgres sqlite chrono" && echo "docs.diesel.rs" > target/doc/CNAME) &&
+  (cd diesel && travis-cargo doc -- --features "postgres sqlite chrono")
+  if [[ "$TRAVIS_RUST_VERSION" == stable ]]; then
+    (echo "docs.diesel.rs" > diesel/target/doc/CNAME)
+  fi &&
```
Comment (killercup) [inline .travis.yml]: Hm, I think this whole docs section should be in `after_success` and only run on stable.

**Comment by killercup** [comment]
LGTM.

Seems to save a bit of time, but not that much. Is the diesel crate built with different features multiple times in the same build (e.g. diesel_test using different features than when building the diesel crate itself)? If so, can we reduce this further?

Totally unscientific timings from travis:

| Test | master | PR branch |
| --- | ---: | ---: |
| nightly sqlite | 8:46 | 7:07 |
| nightly pg | 10:11 | 9:38 |

Long CI builds are luckily not really that much of a problem here, as Travis times out after an hour and one build only takes about 10min right now. (Getting incremental compilation for development will be much more of a difference.)

**Comment by sgrif** [comment]
Yeah it gets built with different features. Shame those don't get cached. I'm less concerned with the time of CI and more with `bin/test`. Maybe we can reorder those so the SQLite tests all happen together and then the PG tests?

**Comment by Bobo1239** [comment]
Just force pushed because I failed the commit message...

**Comment by Bobo1239** [comment]
Reordering the tests in `bin/test` didn't really improve times on my system but I also don't really know how the different crates and features work together so maybe I did it suboptimally.

```
  (cd diesel && cargo test --features "unstable chrono sqlite")
  (cd diesel_cli && cargo test --features "postgres" --no-default-features)
  (cd diesel_tests && cargo test --features "unstable_postgres" --no-default-features)
  (cd diesel_codegen_syntex && cargo test --features "postgres" --no-default-features)
  (cd diesel_cli && cargo test --features "sqlite" --no-default-features)
  (cd diesel_tests && DATABASE_URL=/tmp/test.db cargo test --features "unstable_sqlite" --no-default-features)
  (cd diesel_compile_tests && cargo test)
```

**Comment by killercup** [comment]
FYI, Postgres and SQLite are build by different travis jobs, so that won't matter here. Adding `cache: cargo` to the `.travis.yml` might bring more visible improvements.

@Bobo1239 can you reorder it so that everything with 'sqlite' in it is at the top? That should be as good as we can currently get here.

**Comment by killercup** [comment]
@Bobo1239 Master already adds `cache: cargo`. Can you maybe rebase this PR?

**Comment by Bobo1239** [comment]
yeah I just noticed it

**Comment by killercup** [comment]
I think renaming the examples' crate names is fine. The website links to a
specific commit, so it won't matter when people click on the links over
there; and the meaning should be pretty obvious otherwise.

Thanks for sticking with this PR! I'm sorry for changing so much
cargo/Travis related stuff in the meantime :)

Boris-Chengbiao Zhou <notifications@github.com> schrieb am Sa. 10. Sep.
2016 um 03:19:

**Comment by Bobo1239** [comment]
Examples testing works now but compiletest fails because there are `multiple matching crates for 'diesel_codegen'` in the `target/debug/deps/` directory. Any ideas how to fix this?
(I tried creating a workspace just for the compiletests but cargo complains about diesel_codegen being in another workspace then)

File: .travis.yml [hunk 2]
```diff
@@ -28,7 +27,7 @@ script:
     if [[ "$TRAVIS_RUST_VERSION" == nightly* ]]; then
       (cd examples && ./test_nightly)
     else
-      (cd examples && ./test_stable)
+      (cd examples && TRAVIS_RUST_VERSION=$TRAVIS_RUST_VERSION./test_stable)
```
Comment (killercup) [inline .travis.yml]: Is there a space missing between the variable and the script name? (I'd also explicitly use `env TRAVIS_RUST_…`)

**Comment by killercup** [comment]
Interesting. Full error:

> error: 2 unexpected errors found, 1 expected errors not found
> status: exit code: 101
> command: rustc tests/compile-fail/codegen_does_not_add_save_changes_method_on_structs_without_primary_key.rs -L /tmp --target=x86_64-unknown-linux-gnu --error-format json -Z unstable-options -L /tmp/codegen_does_not_add_save_changes_method_on_structs_without_primary_key.stage-id.compile-fail.libaux -C prefer-dynamic -o /tmp/codegen_does_not_add_save_changes_method_on_structs_without_primary_key.stage-id -L ../target/debug/deps/
> actual errors (from JSON output): [
>     Error {
>         line_num: 2,
>         kind: Some(
>             Error
>         ),
>         msg: "2:11: 2:25: multiple matching crates for `diesel_codegen` [E0464]"
>     },
>     Error {
>         line_num: 2,
>         kind: Some(
>             Note
>         ),
>         msg: "2:11: 2:25: candidates:"
>     },
>     Error {
>         line_num: 2,
>         kind: Some(
>             Note
>         ),
>         msg: "2:11: 2:25: path: /home/travis/build/diesel-rs/diesel/target/debug/deps/libdiesel_codegen.so"
>     },
>     Error {
>         line_num: 2,
>         kind: Some(
>             Note
>         ),
>         msg: "2:11: 2:25: crate name: diesel_codegen"
>     },
>     Error {
>         line_num: 2,
>         kind: Some(
>             Note
>         ),
>         msg: "2:11: 2:25: path: /home/travis/build/diesel-rs/diesel/target/debug/deps/libdiesel_codegen-fd5341f1d057230b.so"
>     },
>     Error {
>         line_num: 2,
>         kind: Some(
>             Note
>         ),
>         msg: "2:11: 2:25: crate name: diesel_codegen"
>     },
>     Error {
>         line_num: 2,
>         kind: Some(
>             Error
>         ),
>         msg: "2:11: 2:25: can\'t find crate for `diesel_codegen` [E0463]"
>     }
> ]

I'm not exactly sure what happens here. It seems that `diesel_codegen` gets compiled a few times (as there are two artifacts available at this point), and `compiletest` uses `prefer-dynamic` in combination with the general deps directory for any libraries.

A few lines above this in the logs, you can see the call

> Running `rustc tests/compile_tests.rs --crate-name compile_tests -g --test -C metadata=6da6d16cb7f217d4 -C extra-filename=-6da6d16cb7f217d4 --out-dir /home/travis/build/diesel-rs/diesel/target/debug --emit=dep-info,link -L dependency=/home/travis/build/diesel-rs/diesel/target/debug/deps --extern diesel_codegen=/home/travis/build/diesel-rs/diesel/target/debug/deps/libdiesel_codegen.so --extern compiletest_rs=/home/travis/build/diesel-rs/diesel/target/debug/deps/libcompiletest_rs-49b8be839ef33737.rlib --extern diesel=/home/travis/build/diesel-rs/diesel/target/debug/deps/libdiesel.rlib -L native=/usr/lib/x86_64-linux-gnu -L native=/usr/lib/x86_64-linux-gnu`

which contains a lot more specific dependencies.

I don't know how to fix this in a nice way. Work arounds might be: Moving the example compile step to be last/moving compile test further up, or manually removing dynamic libs like codegen/codegen_syntex before compiletest.

File: .travis.yml (hunk 2 above)
Comment (Bobo1239) [inline .travis.yml]: Somehow I'm unable to get this right... I tried `[...] && env TRAVIS_RUST_VERSION=$TRAVIS_RUST_VERSION ./test_stable` but the variable is lost in the script. (even when only adding a space `[...] && TRAVIS_RUST_VERSION=$TRAVIS_RUST_VERSION ./test_stable`)

**Comment by sgrif** [comment]
This sounds like it's worth opening a bug with compiletest over?

**Comment by sgrif** [comment]
Or just removing compiletest from the workspace

**Comment by Bobo1239** [comment]
@sgrif Removing doesn't work because of
```
error: current package believes it's in a workspace when it's not:
current:   /home/bobo1239/development/Rust/diesel/diesel_compile_tests/Cargo.toml
workspace: /home/bobo1239/development/Rust/diesel/Cargo.toml
```
or rather
```
error: package `/home/bobo1239/development/Rust/diesel/diesel/Cargo.toml` is a member of the wrong workspace
expected: /home/bobo1239/development/Rust/diesel/diesel_compile_tests/Cargo.toml
actual:   /home/bobo1239/development/Rust/diesel/Cargo.toml
```
after a separate workspace is introduced for the compiletests.

I'll try to somehow find out which files are present in `target/debug/deps` and then think of an appropriate course of action.

**Comment by Bobo1239** [comment]
@killercup Any idea why the version with `env` doesn't work?

**Comment by killercup** [comment]
@Bobo1239 I don't think this does anything on stable: `TRAVIS_RUST_VERSION=$TRAVIS_RUST_VERSION./test_stable` does not execute anything unless you add a space between `$TRAVIS_RUST_VERSION` and `./test_stable`. The travis log [here](https://travis-ci.org/diesel-rs/diesel/jobs/162598788) for stable with postgres doesn't seem to show anything related to compiling the examples.

**Comment by Bobo1239** [comment]
@killercup That makes sense though Rust 1.12 is out so we don't need the conditional anymore anyways.

I just saw the Macro 1.1 PR and will probably postpone updating this until it landed.

**Comment by Bobo1239** [comment]
Hitting the `multiple matching crates for diesel` problem again... I excluded the compile tests from the workspace as I don't know any way to fix this other than manually deleting the lib files which is imo too hacky.

**Comment by sgrif** [comment]
It looks like cargo has improved its caching without workspaces, so there's not as much specific benefit from doing this anymore.

**Comment by sgrif** [comment]
@killercup asked me to take a second look at this.
Super unscientific single run of `rm -rf **/target && time bin/test`:

With PR:
5:31

Without PR:
6:43

Seems good! I will fix the test failures manually

**Comment by sgrif** [comment]
Merged in 9cb56fec2fd7fbf644798b2cc6423b92cb5603a7

**Pull request closed without merging**